astropy/utils/parsing.py

```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Wrappers for PLY to provide thread safety.
+"""
+
+_TAB_HEADER = """# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# This file was automatically generated from ply. To re-generate this file,
+# remove it from this folder, then build astropy and run the tests in-place:
+#
+#   python setup.py build_ext --inplace
+#   pytest {package}
+#
+# You can then commit the changes to this file.
+
+"""
+
+import contextlib
+import functools
+import re
+import os
+import threading
+
+
+_LOCK = threading.RLock()
+
+
+def _add_tab_header(filename, package):
+    with open(filename, 'r') as f:
+        contents = f.read()
+
+    with open(filename, 'w') as f:
+        f.write(_TAB_HEADER.format(package=package))
+        f.write(contents)
+
+
+@contextlib.contextmanager
+def _patch_get_caller_module_dict(module):
+    """Temporarily replace the module's get_caller_module_dict.
+
+    It's wrapped to offset the lookup depth by one level, so that we get the
```

Comment:
Maybe expand a bit on the explanation? I had to look at the `ply` source code to understand why this was needed. Something along the lines of:
```
Inside ``ply.lex`` and ``ply.yacc``, this function is used to get the symbols
that define the parser, by looking at a different level. Here, we wrap the function
to account for the fact that the level changes because of our wrapping.

Note that we cannot avoid ``get_caller_module_dict`` by passing in a ``module``
directly to ``lex`` and ``yacc``, since our symbols are defined inside class methods.
```

Reply:
Done - see what you think now.

Reply:
Yes, a nice balance, thanks!